compiler/rustc_passes/src/liveness.rs

```diff
@@ -470,96 +470,139 @@ impl<'tcx> Visitor<'tcx> for IrMaps<'tcx> {
 
 #[derive(Clone, Copy)]
 struct RWU {
-    reader: Option<LiveNode>,
-    writer: Option<LiveNode>,
+    reader: bool,
+    writer: bool,
     used: bool,
 }
 
-/// Conceptually, this is like a `Vec<RWU>`. But the number of `RWU`s can get
-/// very large, so it uses a more compact representation that takes advantage
-/// of the fact that when the number of `RWU`s is large, most of them have an
-/// invalid reader and an invalid writer.
+/// Conceptually, this is like a `Vec<Vec<RWU>>`. But the number of
+/// RWU`s can get very large, so it uses a more compact representation.
 struct RWUTable {
-    /// Each entry in `packed_rwus` is either INV_INV_FALSE, INV_INV_TRUE, or
-    /// an index into `unpacked_rwus`. In the common cases, this compacts the
-    /// 65 bits of data into 32; in the uncommon cases, it expands the 65 bits
-    /// in 96.
+    /// Total number of live nodes.
+    live_nodes: usize,
+    /// Total number of variables.
+    vars: usize,
+
+    /// A compressed representation of `RWU`s.
     ///
-    /// More compact representations are possible -- e.g., use only 2 bits per
-    /// packed `RWU` and make the secondary table a HashMap that maps from
-    /// indices to `RWU`s -- but this one strikes a good balance between size
-    /// and speed.
-    packed_rwus: Vec<u32>,
-    unpacked_rwus: Vec<RWU>,
+    /// Each word represents 8 different `RWU`s packed together. Each packed RWU
+    /// is stored in 4 bits: a reader bit, a writer bit, a used bit and a
+    /// padding bit.
```

Comment:
```suggestion
    /// padding bit.
    ///
    /// The data for each live node always starts at a word boundary,
   //// so there may be some unused space in between them.
```

Reply:
Would it make sense to use an even smaller word size of 16 or even 8 bits instead?

This would reduce the possible amount of unused space here :thinking:

Reply:
Strictly from memory usage point of view u8 should be the best choice. I pushed the implementation so we can try it out.